router/pkg/authentication/jwks_token_decoder.go

```diff
@@ -132,8 +201,19 @@ func NewJwksTokenDecoder(ctx context.Context, logger *zap.Logger, configs []JWKS
 	if err != nil {
 		return nil, fmt.Errorf("error initializing JWK: %w", err)
 	}
+	return jwks, nil
+}
 
-	return &jwksTokenDecoder{
-		jwks: jwks,
-	}, nil
+// hasAudience is a common intersection function to check if the token's audiences
+func hasAudience(tokenAudiences, expectedAudiences []string) bool {
+	set := make(map[string]struct{}, len(tokenAudiences))
+	for _, item := range tokenAudiences {
+		set[item] = struct{}{}
+	}
+	for _, item := range expectedAudiences {
+		if _, found := set[item]; found {
+			return true
+		}
+	}
+	return false
 }
```

Comment:
```suggestion
func hasAudience(tokenAudiences, expectedAudiences []string) bool {
	for _, expected := range expectedAudiences {
		for _, token := range tokenAudiences {
			if token == expected {
				return true
			}
		}
	}
	return false
}
```

Reply:
I would prefer to keep is it as it is since the proposed solution is O(N^2)

Reply:
Refactored this further, as per @Noroth's suggestion.